docs/usage/exporting_models.md

```diff
@@ -302,9 +235,47 @@ print(m.model_dump_json())
 #> {"date":"2023-10-28"}
 ```
 
+## Subclass instances for fields of BaseModel, dataclasses, TypedDict
+
+When using fields whose annotations are themselves struct-like types (e.g., `BaseModel` subclasses, dataclasses, etc.),
+the default behavior is to serialize the attribute value as though it was an instance of the annotated type,
+even if it is a subclass. More specifically, only the fields from the _annotated_ type will be included in the
+dumped object:
+
+```py
+from pydantic import BaseModel
+
+
+class InnerModel(BaseModel):
+    x: int
+
+
+class SubInnerModel(InnerModel):
+    y: int
+
+
+class OuterModel(BaseModel):
+    inner: InnerModel
+
+
+m = OuterModel(inner=SubInnerModel(x=1, y=2))
+print(m)
+#> inner=SubInnerModel(x=1, y=2)
+print(m.model_dump())
+#> {'inner': {'x': 1}}
+```
+!!! warning "Migration Warning"
+    This behavior is different from how things worked in Pydantic V1, where we would always include
+    all (subclass) fields when recursively dumping models to dicts. The motivation behind this change in
+    behavior is that it helps ensure that you know precisely which fields could be included when serializing,
+    even if subclasses get passed when instantiating the object. In particular, this can help prevent surprises
+    when adding sensitive information like secrets as fields of subclasses.
+
+TODO: Add and document `SerializeAsAny` type.
```

Comment:
I have docs ready for this. @adriangb you didn't do any work on this (or run into issues trying) did you? If not, I'm happy to quickly throw together a PR adding this along with docs after this gets merged.

Reply:
I may have fixed it in #6007 but I didn’t add a type and export it.